config/kraft/server.properties

```diff
@@ -31,22 +31,24 @@ controller.quorum.voters=1@localhost:9093
 
 ############################# Socket Server Settings #############################
 
-# The address the socket server listens on. It will get the value returned from
-# java.net.InetAddress.getCanonicalHostName() if not configured.
+# The address the socket server listens on.
+# This is required for combined role(i.e. process.roles=broker,controller) node to set the listeners at least for controller
```

Comment:
How about this?

> Combined nodes (i.e. those with `process.roles=broker,controller`) must list the controller listener here at a minimum. If the broker listener is not defined, the default listener will use a host name will be equal to the value of `#java.net.InetAddress.getCanonicalHostName()`, with `PLAINTEXT` listener name, and port 9092.

By the way, I find this approach of allowing a default only for "broker" a bit messy. For a separate patch, but maybe we should require both listeners to be explicitly provided. What do you think?

Reply:
> For a separate patch, but maybe we should require both listeners to be explicitly provided. What do you think?

Agree. I'll file a jira for it later. Thanks.

Reply:
[KAFKA-13667](https://issues.apache.org/jira/browse/KAFKA-13667) is created for this improvement. Thanks.

Reply:
> must list the controller listener here at a minimum.
It seems that we can't only list controller listener here in combined mode because `advertised.listeners` must be a subset of `listeners` and advertised listener can not be empty. so we must always set both controller and broker listeners here.

Reply:
@dengziming , make sense. Could you update your finding in [KAFKA-13667](https://issues.apache.org/jira/browse/KAFKA-13667)? Thanks.